stream-rpc/src/json_rpc.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+
+use diem_types::event::EventKey;
+use std::sync::Arc;
+use storage_interface::DbReader;
+use tokio::task::JoinHandle;
+
+use crate::{
+    connection::Client,
+    errors::JsonRpcError,
+    subscriptions::{subscribe_to_events, subscribe_to_transactions},
+};
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum JsonRpcVersion {
+    #[serde(rename = "2.0")]
+    V2,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum RequestIdentifier {
+    Number(u64),
+    String(Box<str>),
+}
+
+impl std::fmt::Display for RequestIdentifier {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            RequestIdentifier::Number(ref v) => std::fmt::Debug::fmt(v, f),
+            RequestIdentifier::String(ref v) => std::fmt::Debug::fmt(v, f),
+        }
+    }
+}
+
+impl std::hash::Hash for RequestIdentifier {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match self {
+            RequestIdentifier::Number(ref v) => v.hash(state),
+            RequestIdentifier::String(ref v) => v.hash(state),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct JsonRpcRequest {
+    jsonrpc: JsonRpcVersion,
+    #[serde(flatten)]
+    pub method_request: MethodRequest,
+    pub id: RequestIdentifier,
+}
```

Comment:
Man it would be nice to have this unified, but I understand why you didn't. Maybe we can rename this a bit to distinguish it from the JsonRpc types?

Reply:
yeah I'm working on making a generic type, given the contract between both `Method` and `MethodRequest` is basically the same for both

it seems like you can emulate inheritance with a wrapping type that is Deref? I may be very wrong in this but we'll see :-)